src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritHandler.java

```diff
@@ -84,33 +94,104 @@
 
         workQueue = new LinkedBlockingQueue<Work>();
-        workers = new ArrayList<EventThread>(numberOfWorkerThreads);
-        for (int i = 0; i < numberOfWorkerThreads; i++) {
-            EventThread eventThread = createEventThread("Gerrit Worker EventThread_" + i);
-            eventThread.start();
-            workers.add(eventThread);
-        }
+        startQueue();
+    }
+
+
+    /**
+     * Returns a factory for creating worker threads for the receiving queue.
+     * Overwrite when you have special requirements, e.g. security constraints
+     *
+     * @return ThreadFactory
+     */
+    protected ThreadFactory getThreadFactory() {
+        return new ThreadFactory() {
+            private final ThreadFactory parent = Executors.defaultThreadFactory();
+            private final AtomicInteger tid = new AtomicInteger(1);
+
+            @Override
+            public Thread newThread(final Runnable task) {
+              final Thread t = parent.newThread(task);
+              t.setName(THREAD_PREFIX + tid.getAndIncrement());
+              return t;
+            }
+          };
+    }
+
+    /**
+     * Starts the executor if it hasn't started yet, or updates the thread-pool size if it is started.
+     */
+    protected void startQueue() {
+      if (executor == null) {
+          logger.debug("Starting the receiving thread pool.");
+          executor = new ThreadPoolExecutor(
+                  numberOfWorkerThreads,
+                  numberOfWorkerThreads,
+                  threadKeepAliveTime, TimeUnit.SECONDS,
+                  new LinkedBlockingQueue<Runnable>(),
+                  getThreadFactory());
+          executor.allowCoreThreadTimeOut(true);
+          //Start with one thread, and build it up gradually as it needs.
+          executor.prestartCoreThread();
+          logger.info("ReceiveQueue started! Current pool size: {}", executor.getPoolSize());
+      } else {
+          if (executor.getCorePoolSize() < numberOfWorkerThreads) {
+              //If the number has increased we need to set the max first, or we'll get an IllegalArgumentException
+              executor.setMaximumPoolSize(numberOfWorkerThreads);
+              executor.setCorePoolSize(numberOfWorkerThreads);
+          } else if (executor.getCorePoolSize() > numberOfWorkerThreads) {
+              //If the number has decreased we need to set the core first.
+              executor.setCorePoolSize(numberOfWorkerThreads);
+              executor.setMaximumPoolSize(numberOfWorkerThreads);
+          }
+          logger.info("ReceiveQueue running. Current pool size: {}. Current Queue size: {}",
+                  executor.getPoolSize(), getQueueSize());
+          logger.info("Nr of active pool-threads: {}", executor.getActiveCount());
+      }
+    }
+
+    /**
+     * Returns the largest number of threads that have ever simultaneously been in the pool.
+     * @return number of threads
+     */
+    int getLargestPoolSize() {
```

Comment:
why package private?

Reply:
Does it have any value for users to know what is the max poolsize that was encountered? I need this method for testing to be at least package private.

Reply:
ok, in those cases a comment explaining it's there for testability is common :)